test/clientconn_state_transition_test.go

```diff
@@ -656,3 +651,111 @@ func (s) TestStateTransitions_WithRPC_NoResolverUpdate(t *testing.T) {
+		b.logger.Logf("testResolverBuilder: returning build failure")
+		return nil, fmt.Errorf("simulated resolver build failure")
+	}
+	return b.manualR.Build(target, cc, opts)
+}
+
+func (b *testResolverBuilder) Scheme() string {
+	return testResolverBuildFailureScheme
+}
+
+// Tests for state transitions when the resolver initially fails to build.
+func (s) TestStateTransitions_ResolverBuildFailure(t *testing.T) {
+	tests := []struct {
+		name         string
+		exitIdleFunc func(ctx context.Context, cc *grpc.ClientConn) error
+	}{
+		{
+			name: "exitIdleByConnecting",
+			exitIdleFunc: func(_ context.Context, cc *grpc.ClientConn) error {
+				cc.Connect()
+				return nil
+			},
+		},
+		{
+			name: "exitIdleByRPC",
+			exitIdleFunc: func(ctx context.Context, cc *grpc.ClientConn) error {
+				if _, err := testgrpc.NewTestServiceClient(cc).EmptyCall(ctx, &testpb.Empty{}); err != nil {
+					return fmt.Errorf("EmptyCall RPC failed: %v", err)
+				}
+				return nil
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			mr := manual.NewBuilderWithScheme("whatever" + tt.name)
+			backend := stubserver.StartTestService(t, nil)
+			defer backend.Stop()
+			mr.InitialState(resolver.State{Addresses: []resolver.Address{{Addr: backend.Address}}})
+
+			dopts := []grpc.DialOption{
+				grpc.WithTransportCredentials(insecure.NewCredentials()),
+				grpc.WithResolvers(&testResolverBuilder{logger: t, manualR: mr}),
+				grpc.WithIdleTimeout(time.Second),
+			}
+
+			cc, err := grpc.NewClient(testResolverBuildFailureScheme+":///", dopts...)
+			if err != nil {
+				t.Fatalf("Failed to create new client: %v", err)
+			}
+			defer cc.Close()
+
+			// Ensure that the client is in IDLE before connecting.
+			ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+			defer cancel()
+			testutils.AwaitState(ctx, t, cc, connectivity.Idle)
+
+			// Subscribe to state updates.
+			stateCh := make(chan connectivity.State, 1)
+			s := &funcConnectivityStateSubscriber{
+				onMsg: func(s connectivity.State) {
+					select {
+					case stateCh <- s:
+					case <-ctx.Done():
+					}
+				},
+			}
+			internal.SubscribeToConnectivityStateChanges.(func(cc *grpc.ClientConn, s grpcsync.Subscriber) func())(cc, s)
+
+			if state := cc.GetState(); state != connectivity.Idle {
+				t.Fatalf("Expected initial state to be IDLE, got %v", state)
+			}
+
+			cc.Connect()
+			wantStates := []connectivity.State{
+				connectivity.Connecting,       // When channel exits IDLE for the first time.
+				connectivity.TransientFailure, // Resolver build failure.
+				connectivity.Idle,             // After idle timeout.
+				connectivity.Connecting,       // When channel exits IDLE again.
+				connectivity.Ready,            // Successful resolver build and connection to backend.
+			}
+			for _, wantState := range wantStates {
+				waitForState(ctx, t, stateCh, wantState)
+				if wantState == connectivity.Idle {
+					tt.exitIdleFunc(ctx, cc)
```

Comment:
Can we make this test the actual RPC error when we use an RPC to exit idle?

Reply:
Maybe `Unavailable` and then it's treated the same as if the resolver did a `ReportError` immediately instead of failing to build?  But probably we should see what the C++/Java lame/failing channels do before deciding, since that seems like the most equivalent scenario in languages where the resolver can't fail to build -- it just doesn't exist.

Reply:
Switched to `Unavailable`. Also changed the idleness manager code to embed the error, so that the RPC can see the status code.

This does mean that the error returned from `Dial` in this scenario would be a status error. But I guess that's an OK thing to do, unless you feel otherwise. In which case, we can explicitly return a new error here and thereby remove the embedding: https://github.com/grpc/grpc-go/blob/76c67d16f885e590b388032a1113a1581d5d5ed6/clientconn.go#L266

Reply:
Made the changes we discussed offline.